c/misra/src/rules/RULE-7-5/IncorrectlySizedIntegerConstantMacroArgument.ql

```diff
@@ -0,0 +1,46 @@
+/**
+ * @id c/misra/incorrectly-sized-integer-constant-macro-argument
+ * @name RULE-7-5: The argument of an integer constant macro shall have an appropriate size
+ * @description Integer constant macros argument values should be values of a compatible size.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/misra/id/rule-7-5
+ *       correctness
+ *       external/misra/c/2012/amendment3
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.IntegerConstantMacro
+import codingstandards.cpp.Literals
+
+predicate matchesSign(IntegerConstantMacro macro, PossiblyNegativeLiteral literal) {
+  literal.isNegative() implies macro.isSigned()
+}
+
+predicate matchesSize(IntegerConstantMacro macro, PossiblyNegativeLiteral literal) {
+  // Wait for BigInt support to check 64 bit macro types.
+  (macro.getSize() < 64 and matchesSign(macro, literal))
```

Comment:
I think it's confusing that `matchesSize` holds if the macro size is >=64, or if the sign doesn't match - because it's not necessarily true that the size does match. I would therefore rewrite this predicate to:
```codeql
literal.getRawValue() <= macro.maxValue() and
literal.getRawValue() >= macro.minValue()
```
Then modify the `where` clause of the query to add in the extra conditions:
```codeql
    not matchesSign(macro, invoke.getExpr()) and
    explanation = " cannot be negative"
    or
    // Wait for BigInt support to check 64 bit macro types.
    macro.getSize() < 64 and
    // Sign must match, otherwise prefer to report the "cannot be negative" message
    matchesSign(macro, invoke.getExpr())
    not matchesSize(macro, invoke.getExpr()) and
    explanation = " is outside of the allowed range " + macro.getRangeString()
```